sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala

```diff
@@ -2018,15 +2007,29 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
         "tableColumns" -> expected.map(c => s"'${c.name}'").mkString(", "),
         "dataColumns" -> query.output.map(c => s"'${c.name}'").mkString(", ")))
   }
+  
+  def numColumnsMismatchError(
+      operator: LogicalPlan,
+      firstNumColumns: Int,
+      invalidOrdinalNum: Int,
+      invalidNumColumns: Int): Throwable = {
+
+    def ordinalNumber(i: Int): String = i match {
+      case 0 => "first"
+      case 1 => "second"
+      case 2 => "third"
+      case i => s"${i + 1}th"
+    }
 
-  def cannotWriteNotEnoughColumnsToTableError(
-      tableName: String, expected: Seq[Attribute], query: LogicalPlan): Throwable = {
     new AnalysisException(
-      errorClass = "_LEGACY_ERROR_TEMP_1203",
+      errorClass = "NUM_COLUMNS_MISMATCH",
       messageParameters = Map(
-        "tableName" -> tableName,
-        "tableColumns" -> expected.map(c => s"'${c.name}'").mkString(", "),
-        "dataColumns" -> query.output.map(c => s"'${c.name}'").mkString(", ")))
+        "operator" -> toSQLStmt(operator.nodeName),
```

Comment:
The value here may be: INSERTINTOSTATEMENT, APPENDDATA.
It's weird, maybe we should pin it to `INSERT INTO`?

Reply:
Maybe, we should introduce `prettyName()` similar to `prettyName` in `Expression`. @cloud-fan @HyukjinKwon WDYT?

Reply:
> It's weird, maybe we should pin it to INSERT INTO?

So far, let's do that.